tests/test_packet_callbacks.py

```diff
@@ -0,0 +1,105 @@
+import pytest
+import zigpy.types as t
+from .async_mock import MagicMock
+from .conftest import make_ieee
+
+
+def make_packet(src_address=None, **overrides):
+    """Build a packet."""
+    defaults = {
+        "src": src_address
+        or t.AddrModeAddress(addr_mode=t.AddrMode.NWK, address=0x1234),
+        "src_ep": 1,
+        "dst": t.AddrModeAddress(addr_mode=t.AddrMode.NWK, address=0x0000),
+        "dst_ep": 1,
+        "tsn": 123,
+        "profile_id": 0x0104,
+        "cluster_id": 0x0006,
+        "data": t.SerializableBytes(b"test"),
+        "lqi": 255,
+        "rssi": -30,
+    }
+    defaults.update(overrides)
+    return t.ZigbeePacket(**defaults)
+
+
+@pytest.mark.parametrize(
+    "filter_address",
+    [
+        None,
+        t.AddrModeAddress(addr_mode=t.AddrMode.NWK, address=0x1234),
+        t.AddrModeAddress(addr_mode=t.AddrMode.IEEE, address=make_ieee()),
+    ],
+)
+async def test_packet_callback_register_cancel(app, filter_address):
+    """Register and cancel packet callback."""
+    cb = MagicMock()
+    cancel = app.register_packet_callback(filter_address, cb)
+    assert cb in app._packet_callbacks[filter_address]
+    cancel()
+    assert cb not in app._packet_callbacks[filter_address]
+    cancel()
+
+
+@pytest.mark.parametrize(
+    ("src_address", "should_trigger"),
+    [
+        (t.AddrModeAddress(addr_mode=t.AddrMode.NWK, address=0x1234), True),
+        (t.AddrModeAddress(addr_mode=t.AddrMode.NWK, address=0x5678), False),
+    ],
+)
+async def test_packet_callback_address_filter(app, src_address, should_trigger):
+    """Source address match."""
+    filt = t.AddrModeAddress(addr_mode=t.AddrMode.NWK, address=0x1234)
+    cb = MagicMock()
+    app.register_packet_callback(filt, cb)
+    pkt = make_packet(src_address=src_address)
+    app.notify_packet_callbacks(pkt)
+    if should_trigger:
+        cb.assert_called_once_with(pkt)
```

Comment:
We avoid using the `assert_called_...` methods in favor of looking at the mock calls list: `assert mock.mock_calls == [call(...`. Otherwise, a typo would result in the mock `assert_` method erroneously working.